{
  "url": "https://api.github.com/repos/shap/shap/issues/188",
  "repository_url": "https://api.github.com/repos/shap/shap",
  "labels_url": "https://api.github.com/repos/shap/shap/issues/188/labels{/name}",
  "comments_url": "https://api.github.com/repos/shap/shap/issues/188/comments",
  "events_url": "https://api.github.com/repos/shap/shap/issues/188/events",
  "html_url": "https://github.com/shap/shap/issues/188",
  "id": 345573826,
  "node_id": "MDU6SXNzdWUzNDU1NzM4MjY=",
  "number": 188,
  "title": "AssertionError: ReverseV2 is not known to be either linear or a supported non-linearity",
  "user": {
    "login": "makcedward",
    "id": 36614806,
    "node_id": "MDQ6VXNlcjM2NjE0ODA2",
    "avatar_url": "https://avatars.githubusercontent.com/u/36614806?v=4",
    "gravatar_id": "",
    "url": "https://api.github.com/users/makcedward",
    "html_url": "https://github.com/makcedward",
    "followers_url": "https://api.github.com/users/makcedward/followers",
    "following_url": "https://api.github.com/users/makcedward/following{/other_user}",
    "gists_url": "https://api.github.com/users/makcedward/gists{/gist_id}",
    "starred_url": "https://api.github.com/users/makcedward/starred{/owner}{/repo}",
    "subscriptions_url": "https://api.github.com/users/makcedward/subscriptions",
    "organizations_url": "https://api.github.com/users/makcedward/orgs",
    "repos_url": "https://api.github.com/users/makcedward/repos",
    "events_url": "https://api.github.com/users/makcedward/events{/privacy}",
    "received_events_url": "https://api.github.com/users/makcedward/received_events",
    "type": "User",
    "site_admin": false
  },
  "labels": [

  ],
  "state": "closed",
  "locked": false,
  "assignee": null,
  "assignees": [

  ],
  "milestone": null,
  "comments": 8,
  "created_at": "2018-07-29T23:54:55Z",
  "updated_at": "2018-08-15T05:01:14Z",
  "closed_at": "2018-07-30T17:44:24Z",
  "author_association": "NONE",
  "active_lock_reason": null,
  "body": "I built an model by using Keras with Tensorflow backend. However, it thrown an error. So I try to figure out the root cause. \r\n\r\n- Download and rerun the [sample Keras LSTM notebook](https://github.com/slundberg/shap/blob/master/notebooks/deep_explainer/Keras%20LSTM%20for%20IMDB%20Sentiment%20Classification.ipynb) to isolate environment issue. So assuming environment is ok\r\n- Current SHAP still support Keras 2.1.x. Although there is warning that SHAP support Keras >= 2.2.0, it seems that can still use SHAP with Keras 2.1.x by running sample notebook\r\n- Comparing the difference between sample notebook Keras model and my model. The sample one is a Sequential model (<keras.models.Sequential at 0x7fd2d83706a0>) when mine is Keras's engine training model (<keras.engine.training.Model at 0x7f8f6e660cc0>). \r\n\r\nAdded sample code in [github repo](https://github.com/makcedward/nlp/blob/master/sample/nlp-model_interpretation-201808.ipynb). Please kindly advise which part do I miss?\r\n\r\n```\r\nAssertionError                            Traceback (most recent call last)\r\n<ipython-input-10-dba880f8c9e3> in <module>()\r\n----> 1 explainer = shap.DeepExplainer(pipeline.model, encoded_x_train[:10])\r\n      2 shap_values = explainer.shap_values(encoded_x_test[:1])\r\n\r\n/anaconda/envs/py35/lib/python3.5/site-packages/shap/explainers/deep.py in __init__(self, model, data, session)\r\n    170                     assert self.num_variable_inputs(op) == 1, \"GatherV2 can only have one varying input!\"\r\n    171                 else:\r\n--> 172                     assert False, op.type + \" is not known to be either linear or a supported non-linearity!\"\r\n    173 \r\n    174         if not self.multi_output:\r\n\r\nAssertionError: ReverseV2 is not known to be either linear or a supported non-linearity!\r\n```",
  "closed_by": {
    "login": "slundberg",
    "id": 3740613,
    "node_id": "MDQ6VXNlcjM3NDA2MTM=",
    "avatar_url": "https://avatars.githubusercontent.com/u/3740613?v=4",
    "gravatar_id": "",
    "url": "https://api.github.com/users/slundberg",
    "html_url": "https://github.com/slundberg",
    "followers_url": "https://api.github.com/users/slundberg/followers",
    "following_url": "https://api.github.com/users/slundberg/following{/other_user}",
    "gists_url": "https://api.github.com/users/slundberg/gists{/gist_id}",
    "starred_url": "https://api.github.com/users/slundberg/starred{/owner}{/repo}",
    "subscriptions_url": "https://api.github.com/users/slundberg/subscriptions",
    "organizations_url": "https://api.github.com/users/slundberg/orgs",
    "repos_url": "https://api.github.com/users/slundberg/repos",
    "events_url": "https://api.github.com/users/slundberg/events{/privacy}",
    "received_events_url": "https://api.github.com/users/slundberg/received_events",
    "type": "User",
    "site_admin": false
  },
  "reactions": {
    "url": "https://api.github.com/repos/shap/shap/issues/188/reactions",
    "total_count": 0,
    "+1": 0,
    "-1": 0,
    "laugh": 0,
    "hooray": 0,
    "confused": 0,
    "heart": 0,
    "rocket": 0,
    "eyes": 0
  },
  "timeline_url": "https://api.github.com/repos/shap/shap/issues/188/timeline",
  "performed_via_github_app": null,
  "state_reason": "completed"
}
